storage/src/main/java/org/openmbee/sdvc/storage/S3Storage.java

```diff
@@ -0,0 +1,130 @@
+package org.openmbee.sdvc.storage;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import java.io.ByteArrayInputStream;
+import java.util.Date;
+import java.util.Optional;
+import org.apache.tika.mime.MimeTypes;
+import org.openmbee.sdvc.artifacts.storage.ArtifactStorage;
+import org.openmbee.sdvc.json.ElementJson;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class S3Storage implements ArtifactStorage {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private AmazonS3 s3Client;
+
+    @Value("${s3.endpoint}")
+    private String ENDPOINT;
+
+    @Value("${s3.access_key}")
+    private String ACCESS_KEY;
+
+    @Value("${s3.secret_key}")
+    private String SECRET_KEY;
+
+    @Value("${s3.region}")
+    private String REGION;
+
+    @Value("${s3.bucket:#{null}}")
+    private Optional<String> BUCKET;
+
+
+    private MimeTypes mimeTypes = MimeTypes.getDefaultMimeTypes();
+
+    private AmazonS3 getClient() {
+        if (s3Client == null) {
+            AWSCredentials credentials = new BasicAWSCredentials(ACCESS_KEY, SECRET_KEY);
+            ClientConfiguration clientConfiguration = new ClientConfiguration();
+            clientConfiguration.setSignerOverride("AWSS3V4SignerType");
+
+            s3Client = AmazonS3ClientBuilder
+                .standard()
+                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(ENDPOINT, REGION))
+                .withPathStyleAccessEnabled(true)
```

Comment:
does this work with new aws buckets or does it need to be a config?

Reply:
I'm iffy about this as well. S3 supports both virtual-hosted-style and path-style by default, with the only difference being that it requires a region in the endpoint for the pathStyleAccess. By default, it tries to autodetect, in case a bucketname is an invalid dns host, but I thought it would be better to actually know that it is using path style access.

Reply:
can we do a quick test with our internal s3 for a new bucket? it's a bit unclear in the blog post with the update - it originally said any new buckets created after Sep 30, 2020 would require virtual hosted access, but that they're extending the deprecation date (it sounds like they're delaying deprecation for all buckets though)

https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/